### PR TITLE
Part 1 of #9013

### PR DESCRIPTION
**Pull request by lyakh**
a split up of https://github.com/thesofproject/sof/pull/9013/commits/fb6286dbfc48e3f8d12dc99d5ac982cfd4d16124 . Can be merged independently since this code isn't used yet, but to work it'll need the rest of #9013, including a Zephyr update, which is now available as #9019

### CODE REVIEW

**Comment by marc-hb** [comment]
Lots of build failures in https://sof-ci.01.org/sofpr/PR9024/build3859/build/index.html and https://github.com/thesofproject/sof/actions/runs/8633265152/job/23665921289?pr=9024

**Comment by lyakh** [comment]
The fuzzer failure https://github.com/thesofproject/sof/actions/runs/8657935406/job/23740891264?pr=9024 is curious, but I'm rather sure it's unrelated. The whole library management depends on IPC4, and all these run-time changes only affect it. The only potentially relevant for IPC3 change is the change for rimage, which makes it less sensitive to invalid ELF files, but that's a build-time change, and as far as I understand the fuzzer doesn't attempt invalid builds.

**Comment by marc-hb** [comment]
@andyross  I downloaded the input "crash" file (attached again here for convenience), debugged it and noticed something that looks bad.

[IPC3-fuzz9024-logs.zip](https://github.com/thesofproject/sof/files/14964498/IPC3-fuzz9024-logs.zip)

The IPC3 `mailbox_validate()` function returns NULL when the `hdr` is malformed. I'm not sure what the "real" IPC3 code does with that NULL (I didn't look) but what you did in commit c50eddce1f81b and `src/platform/posix/ipc.c` does not look correct: you keep going like nothing happened?!

Most of the time it surprisingly does not crash (lack of defensive programming?) but it's not too surprising if that crashes from time to time - as it just did in https://github.com/thesofproject/sof/actions/runs/8657935406/job/23740891264?pr=9024

**Comment by marc-hb** [comment]
>  I'm not sure what the "real" IPC3 code does with that NULL (I didn't look) 

Actually, AMD, IMX and mediatek keep trying to process that NULL like posix/ipc.c does. Not sure about Intel

**Comment by andyross** [comment]
To clarify: you're seeing a NULL header passed to ipc_cmd() because of an invalid parse (an expected failure, this is fuzzing aftera all) that happens in mailbox_validate()?   I'll admit I'm not sure exactly what the specified behavior there is.   The POSIX code was written to try to duplicate the original scheme, but... yeah, it's pretty confusing how this is intended to work.  There's a big comment in posix/ipc.c trying to explain what I think is happening.

FWIW, src/ipc/ipc-zephyr.c looks broadly similar: I see ipc_compact_read_msg() returning an uninspected outptut of mailbox_validate(), which ends up in ipc_platform_do_cmd() where it's passed straight to ipc_cmd().  Basically the same code path?

My gut says this is a core issue and that ipc_cmd() is expected to be tolerant of NULLs at runtime, I guess?  But we could absolutely disallow it and check before the call too.

**Comment by marc-hb** [comment]
> My gut says this is a core issue and that ipc_cmd() is expected to be tolerant of NULLs at runtime, I guess?

I think you must be right.

Now I suspect the timeout logic in the fuzzing framework. I suspect it because the failing run lasted exactly as long as the passing runs.

I ran it a second time and it passed https://github.com/thesofproject/sof/actions/runs/8657935406/job/23776321942

I also run the crash input locally a million times and no problem were ever found.

Let's ignore this. We still have the .zip attached if we ever need to get back to it.

**Review by lgirdwood — APPROVED** [review]

File: src/samples/audio/smart_amp_llext/CMakeLists.txt [hunk 1]
```diff
@@ -53,7 +53,16 @@ target_compile_options(${MODULE} PRIVATE
 	-save-temps -O2
 )
 
+if("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "zephyr")
```
Comment (marc-hb) [inline src/samples/audio/smart_amp_llext/CMakeLists.txt:56]: Add a "FIXME: replace this hack with new `add_llext_target()` in Zephyr"

File: src/library_manager/llext_manager.c [hunk 2]
```diff
@@ -170,7 +171,7 @@ static int llext_manager_link(struct sof_man_fw_desc *desc, struct sof_man_modul
 			      uint32_t module_id, struct module_data *md, const void **buildinfo,
 			      const struct sof_man_module_manifest **mod_manifest)
 {
-	size_t mod_size = desc->header.preload_page_count * PAGE_SZ;
+	size_t mod_size = desc->header.preload_page_count * PAGE_SZ - 0x8000;
```
Comment (marc-hb) [inline src/library_manager/llext_manager.c:174]: Magic number

File: scripts/llext_link_helper.py [hunk 3]
```diff
@@ -33,21 +35,53 @@ def main():
...
 
-	upper = data_offset - rodata_offset + rodata_addr + 0xf
-	data_addr = upper - (upper % 0x10)
+	command = [args.command]
+
+	for i in range(len(sections)):
+		try:
+			offset = elf.get_section_by_name(sections[i]).header.sh_offset
+			size = elf.get_section_by_name(sections[i]).header.sh_size
+		except:
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:61]: No, just catch the exception you're expecting. Copy/paste its name from the stack trace, it should take seconds. Also, use `pylint`.

https://pylint.pycqa.org/en/latest/user_guide/messages/warning/broad-exception-caught.html

File: scripts/llext_link_helper.py [hunk 4]
```diff
@@ -7,9 +7,11 @@
 # portable way to do that. Therefore we pass the linker path and all the command
 # line parameters to this script and call the linker directly.
 
+import os
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:10]: Where is this used in this commit?

**Review by marc-hb — CHANGES_REQUESTED** [review]
https://pylint.pycqa.org/en/latest/user_guide/messages/warning/broad-except.html

**Comment by lgirdwood** [comment]
> https://pylint.pycqa.org/en/latest/user_guide/messages/warning/broad-except.html

@marc-hb these are all minor improvements, the main benefit is getting this used and @lyakh can circle back and do the updates when he's back from vacation.  I'm going to merge so we can start looking at the next steps for testing, CI and signing.

**Review by kv2019i — APPROVED** [review]
The lllext_link_helpe.py points are valid, but otherwise this is ready to merge.

**Comment by marc-hb** [comment]
Actually, "broad-except" is not a minor warning and this is the type of coding mistakes that gets fixed only if/when being hit hard by it and wasting a lot of time. Same as ignoring errors in C.

**Pull request merged**